// testcafe/lib/browser/provider/built-in/dedicated/chrome/cdp-client/shared-adapter-initializer.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const execution_context_1 = __importDefault(require("./execution-context"));
const clients_manager_1 = require("./clients-manager");
const adapter_1 = require("../../../../../../shared/adapter");
const domUtils = __importStar(require("./utils/dom-utils"));
const positionUtils = __importStar(require("./utils/position-utils"));
const styleUtils = __importStar(require("./utils/style-utils"));
const eventUtils = __importStar(require("./utils/event-utils"));
const create_event_sequence_1 = __importDefault(require("./utils/create-event-sequence"));
adapter_1.initializeAdapter({
    PromiseCtor: Promise,
    nativeMethods: {
        setTimeout,
        clearTimeout,
        arrayIndexOf: Array.prototype.indexOf,
        arraySplice: Array.prototype.splice,
        arraySlice: Array.prototype.slice,
        arrayFilter: Array.prototype.filter,
        objectAssign: Object.assign,
        objectKeys: Object.keys,
        dateNow: Date.now,
    },
    scroll: async (el, opts) => {
        let currCxt = execution_context_1.default.current;
        let result = null;
        let margin = void 0;
        do {
            const { exceptionDetails, result: resultObj } = await clients_manager_1.getClient().Runtime.callFunctionOn({
                returnByValue: true,
                awaitPromise: true,
                executionContextId: execution_context_1.default.getCurrentContextId(),
                arguments: [{ objectId: el.objectId }, { value: opts }, { value: margin }],
                functionDeclaration: `function (el, opts) {
                    return window["%proxyless%"].scroll(el, opts);
                }`,
            });
            if (exceptionDetails)
                throw exceptionDetails;
            const scrollResult = resultObj.value;
            if (currCxt && currCxt !== currCxt.parent) {
                // TODO:
                //el           = findIframeByWindow(currCxt);
                currCxt = currCxt.parent;
                result = result !== null && result !== void 0 ? result : scrollResult.scrollWasPerformed;
                margin = scrollResult.maxScrollMargin;
                opts.offsetX = scrollResult.offsetX;
                opts.offsetY = scrollResult.offsetY;
            }
        } while (currCxt && currCxt !== currCxt.parent);
        return result;
    },
    browser: { isChrome: true },
    featureDetection: {
        isTouchDevice: false,
    },
    createEventSequence: create_event_sequence_1.default,
    sendRequestToFrame: () => { },
    getElementExceptUI: positionUtils.getElementFromPoint,
    dom: domUtils,
    position: positionUtils,
    style: styleUtils,
    event: eventUtils,
});
